node/src/storage_mngr/node_migrations.rs

```diff
@@ -0,0 +1,367 @@
+use super::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use witnet_data_structures::chain::{ChainState, Epoch};
+
+macro_rules! as_failure {
+    ($e:expr) => {
+        failure::Error::from_boxed_compat(Box::new($e))
+    };
+}
+
+/// Return the version of the `ChainState` serialization. Returns error on end of file.
+fn check_chain_state_version(chain_state_bytes: &[u8]) -> Result<u32, ()> {
+    if chain_state_bytes.is_empty() {
+        return Err(());
+    }
+
+    // Before versioning support, the first byte of the serialization of ChainState was the tag of
+    // an Option, which is one byte that must be either 0 or 1.
+    if chain_state_bytes[0] == 0 || chain_state_bytes[0] == 1 {
+        Ok(0)
+    } else {
+        // After versioning support, there is a db_version before the serialization.
+        // This field is a u32 (little endian) so it takes the first 4 bytes.
+        // db_version % 256 must never be 0 or 1, because that can be confused with version 0.
```

Comment:
This is brilliant :)

Reply:
Simply solves the one concerned I always had about bincode migrations: telling a versioned structure from the original one that may had virtually anything as its first byte. Luckily, the fact that in the case of the chain state this is an enum, this will totally do the trick!